Use CakeBuild for building osu!

miterosan opened the pull request:
Use cakebuild for building, testing and analyzing the osu! source code.

This allows running the code checks locally.

~Please note that the code file sanity tool and NVika do not have bindings for cakebuild yet.~

miterosan committed:
Create an inital cake buildscript

miterosan committed:
Allow testing

miterosan committed:
Prepare inspect code

miterosan committed:
Also analyze the code for spaces and warnings.

miterosan committed:
Restore nuget before analyzing the binaries.

VPeruS commented on build.cake:
```diff
@@ -0,0 +1,69 @@
+///////////////////////////////////////////////////////////////////////////////
+
+var target = Argument("target", "Build");
+var framework = Argument("framework", "net471");
```
Desktop has netcoreapp2.1 as target framework

miterosan replied:
This is true, but it was always build with net471 in the appveyor.xml. @peppy can we switch over to netcoreapp2.1 or do we stay on net471 for now?

VPeruS replied:
appveyor use value from csproj -> netcoreapp2.1

miterosan replied:
nevermind then

peppy replied:
We made the complete switch yesterday via https://github.com/ppy/osu/pull/3061

miterosan committed:
Merge branch 'master' of https://github.com/ppy/osu into Cakefy_osu!

# Conflicts:
#	appveyor.yml

miterosan committed:
Build against netcoreapp2.1 instead of net471

swoolcock commented:
Will this work for Xamarin?

peppy commented:
It's just a script engine; it can be made to work with anything.

peppy commented:
Can you update this to consume CFS via nuget? See https://github.com/ppy/osu/pull/3137

miterosan commented:
.-.  I am against having the end user installing codefilesanity globally on his pc. The local build should not require random programs preinstalled. The point of switching to cakebuild was that the developers can test and validate the code by just running build.ps1/build.sh. Installing CodeFileSanity as a global tool would defeat that purpose. 

Dotnetcoretools are supposed to be added as a packagereference and then called over dotnet somedotnettool osu.desktop.csproj. You can not use a dotnettool without 

The way we can go about this is implementing the cake addin alias interface. This allows customizability of codefilesanity inside the cake build script.
https://codeopinion.com/how-to-create-cake-addin/

TLDR:
- dotnetcore Global tools require installing
- dotnetcore tools can only appliend on single csproj files that packagereference the tool.
- implementing the cakealias interface is clean and customizable.

^ I am currently on this.

peppy commented:
Oh yeah, no shit. I am 100% with you on that one. I even explained to smoogi the same thing over a good 30 minutes. The global install is fine in the appveyor config because it's a once-off run, but doing that locally would be blasphemy.

miterosan committed:
Use CodeFileSanity as a cake addin.

miterosan committed:
Move compile to the correct task.

The dotnetcore... commands all contain compiling.

miterosan committed:
:V JetBrains.ReSharper.CommandLineTools is an addin and not a tool

miterosan committed:
Merge branch 'master' of https://github.com/ppy/osu into Cakefy_osu!

# Conflicts:
#	appveyor.yml

miterosan commented:
This build needs to be repeated when https://github.com/peppy/CodeFileSanity/pull/5 is merged and deployed.

miterosan committed:
Use coreclr instead of .framework.

miterosan committed:
Use netcore cake

miterosan committed:
Linux users can use powershell on, linux

miterosan committed:
Allow running inspect core with cakeclr.

Sadly window only because nvika and resharper cmd tools require net45.

miterosan committed:
Add inspectcodereport to the gitignore

miterosan committed:
remove unfnished comment.

miterosan committed:
Allow running on mono if possible. (non windows)

miterosan committed:
Add cake.csproj

miterosan committed:
Use the correct tool executable. Better mono detection

miterosan committed:
rm bad sign

miterosan committed:
silent the test for mono

miterosan committed:
minus minus

miterosan committed:
adjust the path to the commandline tools

miterosan committed:
Revert "Allow running on mono if possible. (non windows)"

https://youtrack.jetbrains.com/issue/RSRP-410004

This reverts commit 580d478dc80bafbbe39eede20ff1e9f6dc234dc6.

peppy commented on build.cake:
```diff
@@ -38,17 +38,14 @@ Task("Test")
 .IsDependentOn("Compile")
 .Does(() => {
     var nVikaToolPath = GetFiles("./tools/NVika.MSBuild.*/tools/NVika.exe").First();
     var nugetToolPath = GetFiles("./tools/NuGet.CommandLine.*/tools/NuGet.exe").First();
```
`dotnet build` automatically does a nuget restore. this is unnecessary.

miterosan replied:
The tests are not restored at that point of time (this runs before the tests). Windows nuget makes no sense here, yes.

miterosan committed:
Remove the usage of windows nuget restore and replace it with dotnet restore

peppy commented:
@miterosan what's the status on this one?

miterosan commented:
@peppy Waiting that you merge/review peppy/CodeFileSanity#5. This and the codeFileSanity are ready on my side.

peppy commented:
oh, sorry. thought i already got that one in...

miterosan commented:
Bout this issue. This should be ready. Please note that you need to create a release of the codefilesanity one for this to make sense.

peppy commented:
Should already be published: https://www.nuget.org/packages/CodeFileSanity/0.0.21

miterosan committed:
Pin the addin and tool versions down.

miterosan committed:
Merge branch 'master' of https://github.com/ppy/osu into Cakefy_osu!

miterosan commented:
I pinned the tool and addin versions down to allow repeatable builds.
The is PR is mergeable from my side when the CI goes green.

miterosan committed:
Remove Redundant parentheses

miterosan committed:
Report the test results to appveyor

miterosan commented:
for reference: why is the build green if there are failing tests?

peppy commented:
The exit code is probably 0?

miterosan committed:
Use Dotnet core VSTest

This allows only running one testrun.

miterosan committed:
Reduce the verbosity of the Restore.

miterosan committed:
Build the projects before testing

miterosan committed:
Seperate the compile from restoring.

miterosan committed:
Merge branch 'master' into Cakefy_osu!

miterosan committed:
Merge branch 'master' into Cakefy_osu!

miterosan commented:
So... this should be ready for review again.

peppy committed:
Merge branch 'master' into Cakefy_osu!

miterosan committed:
Merge branch 'master' into Cakefy_osu!

miterosan committed:
Minify the buildscript and update to cake.0.30.0

Also remove the dependency on Cake.CoreCLR
cake.tool does not use the packages.config.

miterosan committed:
Also rename the cake project in the bootstrap script

miterosan committed:
Merge branch 'master' into Cakefy_osu!

miterosan commented:
Is ready for review.

I increased the used cakebuild version to the latest.
Tested the buildscript on both windows 1809 and on lubuntu (Ubuntu 18.04.1 LTS x84-64, linux powershell version v6.1.0-preview.2, dotnet version 2.1.302)

jai-x commented:
I would suggest a squash and rebase into one commit with a properly written commit message.

miterosan commented:
can do when approval is given

smoogipoo commented:
On macOS: `Error: Failed to install tool 'NVika.MSBuild'.`

miterosan committed:
Add config file

miterosan committed:
Use the coreclr instead of caketool, hoping that the version fixes running it on linux

miterosan committed:
Merge branch 'Cakefy_osu!' of https://github.com/miterosan/osu into Cakefy_osu!

miterosan committed:
Merge branch 'master' into Cakefy_osu!

peppy committed:
Merge branch 'master' into Cakefy_osu!

miterosan committed:
Merge branch 'master' into Cakefy_osu!

miterosan committed:
Merge branch 'master' into Cakefy_osu!

miterosan committed:
Merge branch 'master' into Cakefy_osu!

miterosan committed:
Merge branch 'master' into Cakefy_osu!

peppy commented:
The readme should probably be updated with build/test instructions using this method.

miterosan committed:
Reference the build script in the README.

miterosan committed:
Merge branch 'Cakefy_osu!' of https://github.com/miterosan/osu into Cakefy_osu!

peppy committed:
Merge branch 'master' into Cakefy_osu!

peppy commented:
Is there a reason for only providing a powershell bootstrap for this? May be useful to also have a bash script.

peppy commented:
```PS C:\Users\Dean\Projects\osu> .\build.ps1
.\build.ps1 : File C:\Users\Dean\Projects\osu\build.ps1 cannot be loaded because running scripts is disabled on this
system. For more information, see about_Execution_Policies at https:/go.microsoft.com/fwlink/?LinkID=135170.
At line:1 char:1
+ .\build.ps1
+ ~~~~~~~~~~~
    + CategoryInfo          : SecurityError: (:) [], PSSecurityException
    + FullyQualifiedErrorId : UnauthorizedAccess```

is this normal for a vanilla windows install?

goatdavid commented:
yeah, iirc you need to run powershell as admin and use Set-ExecutionPolicy
RemoteSigned

2018年10月29日(月) 午前10:37 Dean Herbert <notifications@github.com>:

> .\build.ps1 : File C:\Users\Dean\Projects\osu\build.ps1 cannot be loaded because running scripts is disabled on thissystem. For more information, see about_Execution_Policies at https:/go.microsoft.com/fwlink/?LinkID=135170.At line:1 char:1+ .\build.ps1+ ~~~~~~~~~~~
>     + CategoryInfo          : SecurityError: (:) [], PSSecurityException
>     + FullyQualifiedErrorId : UnauthorizedAccess```
> is this normal for a vanilla windows install?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ppy/osu/pull/3135#issuecomment-433934355>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AYw4ZAqJ4qViQ4yqhFhbJkjRtPTP17hRks5upxK5gaJpZM4Vln7H>
> .
>

miterosan committed:
Create a simple bootstrapper in the bash script language

miterosan committed:
Mention the bash script in the readme.

miterosan committed:
Merge branch 'Cakefy_osu!' of https://github.com/miterosan/osu into Cakefy_osu!

miterosan committed:
Remove the framework settings. It is not used.

peppy committed:
Fix weird code formatting

peppy committed:
Reword readme

peppy committed:
Merge branch 'master' into Cakefy_osu!

peppy commented:
Works as expected on mac and windows. Let's give this a try.
